geo: return 3D+ dimensions from GEOS

otan opened the pull request:
Currently any dimension about 2 was omitted with GEOS returned. Fix this
by telling the EWKB reader to return 3 if there is a Z coordinate.

Note this does not work for EMPTY geometries -- but this should be fine
since we don't store 3D/4D objects at this time.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/53349)
<!-- Reviewable:end -->

sumeerbhola reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53349)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan and @sumeerbhola)

___
*[pkg/geo/parse_test.go, line 250 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/53349#-MFX0tXTBczY3YVGy6m7:-MFX0tXTBczY3YVGy6m8:b1vbtx8) ([raw file](https://github.com/cockroachdb/cockroach/blob/91f573ffb64d750efee7af46b0acd2aea7f7f51b/pkg/geo/parse_test.go#L250)):*
> ```Go
> 		{"POINT Z (1 2 3)", "only 2D objects are currently supported"},
> 		{"POINT M (1 2 3)", "only 2D objects are currently supported"},
> 		{"POINT ZM (1 2 3)", "only 2D objects are currently supported"},
> ```

are these errors a change in behavior?
___
*[pkg/geo/geos/geos.cc, line 415 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/53349#-MFX17cE2GyGQvZzS5HE:-MFX17cE2GyGQvZzS5HF:b7k6d0s) ([raw file](https://github.com/cockroachdb/cockroach/blob/91f573ffb64d750efee7af46b0acd2aea7f7f51b/pkg/geo/geos/geos.cc#L415)):*
> ```cpp
>   lib->GEOSWKBWriter_setByteOrder_r(handle, wkbWriter, 1);
>   if (hasZ) {
>     lib->GEOSWKBWriter_setOutputDimension_r(handle, wkbWriter, 3);
> ```

is there a test that shows that Z is preserved for geom?
___
*[pkg/sql/logictest/testdata/logic_test/geospatial, line 4034 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/53349#-MFX0c3QC3u45PF3Ox1d:-MFX0c3QC3u45PF3Ox1e:b9zpqnx) ([raw file](https://github.com/cockroachdb/cockroach/blob/516ce1a9d8b212302b96884d17a4ddedd552881d/pkg/sql/logictest/testdata/logic_test/geospatial#L4034)):*
> ```
> 
> query T
> SELECT ST_Summary('POINT(0 0 0 0)'::geometry)
> ```

why were these tests removed?


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53349)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @sumeerbhola)

___
*[pkg/geo/parse_test.go, line 250 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/53349#-MFX0tXTBczY3YVGy6m7:-MFX1RIv7lyRQrSCV8oa:bsxa15i) ([raw file](https://github.com/cockroachdb/cockroach/blob/91f573ffb64d750efee7af46b0acd2aea7f7f51b/pkg/geo/parse_test.go#L250)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

are these errors a change in behavior?
</blockquote></details>

Yes. Previously they returned only their 2D state.
We explicitly error these out here: https://github.com/cockroachdb/cockroach/blob/master/pkg/geo/geo.go#L807.
The M state is also not preserved with GEOS' WKT parsing, so we'd need to rewrite this at some later point (tracking issue https://github.com/cockroachdb/cockroach/issues/53091)
___
*[pkg/geo/geos/geos.cc, line 415 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/53349#-MFX17cE2GyGQvZzS5HE:-MFX1utbEHFgT1B-6MQu:besi1hx) ([raw file](https://github.com/cockroachdb/cockroach/blob/91f573ffb64d750efee7af46b0acd2aea7f7f51b/pkg/geo/geos/geos.cc#L415)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

is there a test that shows that Z is preserved for geom?
</blockquote></details>

The error case is intended to catch these (for now). When we support Z properly, there will need to be more.
___
*[pkg/sql/logictest/testdata/logic_test/geospatial, line 4034 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/53349#-MFX0c3QC3u45PF3Ox1d:-MFX1Tzv5ZT3dE-q5Hzf:b-mnkfyu) ([raw file](https://github.com/cockroachdb/cockroach/blob/516ce1a9d8b212302b96884d17a4ddedd552881d/pkg/sql/logictest/testdata/logic_test/geospatial#L4034)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

why were these tests removed?
</blockquote></details>

They error now -- so I have removed them.


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (approved):
> Currently any dimension about 2 was omitted with GEOS returned

did you mean "... above 2 was omitted when GEOS ..."?

<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 3 of 3 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53349#-:-MFXKceSDK1gl6vScvRm:b-r47djq)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @sumeerbhola)



<!-- Sent from Reviewable.io -->

otan committed:
geo: return 3D+ dimensions from GEOS

Currently any dimension above 2 was omitted with GEOS. Fix this
by telling the EWKB reader to return 3 if there is a Z coordinate.

Note this does not work for EMPTY geometries -- but this should be fine
since we don't store 3D/4D objects at this time.

Release note: None

otan commented:
bors r=sumeerbhola
